### ISSUE TITLE
Identical Appointment on both Calendars deleted and recreated at sync

### ISSUE THREAD

**Issue by charlestek** [phw198/OutlookGoogleCalendarSync#568]
**OGCS Version**: v2.7.5
**Installed or Portable**: Installed

**Problem Description**  
Appointment with same ascii name at same time 
created independently on both Google Calendar and
Outlook Calendar is deleted and recreated on Google calendar
after sync.

**Steps to Reproduce the Issue**  
1. Create Appointment in Google Calendar
with subject:  "Test Calendar" at mm/dd/yyyy in the future
All other fields blank.
2. Create Appointment in Outlook Calendar with 
subject:  "Test Calendar" at same mm/dd/yyyy in the future

3. Sync manually or automatically.
Settings Outlook <-> Google only "Merge with existing entries" is checked.

4.   Result, another appointment named "Test Calendar" is created on Google Calendar at same mm/dd/yyyy, but apparently
it is deleted first from Google Calendar then recreated on Google Calendar, so there remains one entry  on the Google Calendar.

5.  If an appointment at the identical time and identical subjects with all other fields blank exists in both calendar's should not your algorithm decide first that they are identical (not sure whether Google and Outlook fields have extra trailing spaces or nulls) and then not bother to delete/recreate on top of the identical appointment (Identical per ascii title) ?????

**Comment by phw198** [comment]
I can't actually replicate this as you describe in v2.7.5, but I had some changes coming through around this for the next version anyway. I'll provide a hotfix and see if that solves it for you...

**Comment by phw198** [comment]
:package: As promised, improvements merged in for [hotfix v2.7.5.5.zip](https://github.com/phw198/OutlookGoogleCalendarSync/files/2299319/v2.7.5.5.zip) ([see instructions for applying](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-Problems#applying-a-hotfix)).

However, with your recent update above, I don't believe this will actually fix it for you....further post to follow.

**Comment by charlestek** [comment]
It appears that the program is working fine without duplicates, but I will try the hotfix in the afternoon.  I have to go to bed and get some sleep.
 
Thank you so much for your help.

**Comment by phw198** [comment]
> I also had the program installed on outlook on two different desktop pc's so I wound up having a zillion duplicates

Ahhhhh, _now_ you're saying the true set up. That's not a supported configuration, but one I'm actively working on via Issue #382 and hope to have a new alpha release out this weekend.

> Outlook -> Google, Outlook <-Google and Outlook<->Good deletion checkboxes in all 3 use cases, I still don't understand it, and your help only describes the function and does not explain anything at all.
I would encourage to simplify this functionality for a lay person.   Triple functionality drop down boxes in the winform is great for elegance, but not so much for intuitiveness......<snip>Outlook <-> Google should be completely removed

And I have no clue what you're talking about either :confused: It allows the user to chose a one-way sync in either direction, or a two-way sync. That's the top two options in the below graphic. Removing core functionality is quite simply not going to happen. 

However, I think you're getting  confused with the setup of OGCS being installed on two different PCs and, I assume, syncing to a common Google calendar - which is the unsupported first item below with a red cross next to it.

<img src="https://github.com/phw198/OutlookGoogleCalendarSync/raw/master/docs/images/supported_sync_configs.png" width="150"/>

> Also, I'm now using the latest version from github, where before I used an older non alpha, and changed the sync frequency from the default to every 15minutes.   Then two weeks later after missing some doctor's appointments,  I realized that the program didn't do a sync at all for two weeks , that somehow I think changing the time interval caused a bug so that the program did not get triggered at all.

Now that might indeed be a new bug that no one's reported before. As there is no "default" sync frequency (it is actually turned off), do you recall what you had configured on the beta release before upgrading to the alpha?

**Comment by charlestek** [comment]
Outlook calendar sync is only on one PC now.  But, I just had exited from outlook, Calendar sync failed to sync.
 
Restarted Outlook, did a manual sync and it now deleted 61 valid Outlook appointments.
 
I have to uninstall your program, I cannot afford to have my calendar messed up like this.....
  
Sync version: 2.7.5.0
Sync started at 8/18/2018 6:28:53 PM
Syncing from 7/18/2018 to 10/18/2018
Outlook ↔ Google 
Finding Calendar Entries
Scanning Outlook calendar... 
66 Outlook calendar entries found. 
Scanning Google calendar... 
3 Google calendar entries found. 
Total inc. recurring items spanning sync date range...
Outlook 64, Google 3  
0 Google calendar entries to be deleted. 
0 Google calendar entries to be created. 
3 calendar entries to be compared. 
Comparing 3 existing Google calendar entries
0 entries updated.  
61 Outlook calendar entries to be deleted. 
0 Outlook calendar entries to be created. 
3 calendar entries to be compared.
 
Deleting 61 Outlook calendar entries
(deleted labels for privacy)
Comparing 3 existing Outlook calendar entries
0 entries updated.
 
Sync finished with success!

**Comment by charlestek** [comment]
Deleted the 61 appointments in Google Calendar as well.   Now I have to manually re-enter all of them.

**Comment by phw198** [comment]
Deleted items will be in your deleted items folder in Outlook, the Bin in Google, so should be straight forward to restore.

Unfortunately, as I said, it's unsupported in the configuration you had and two way sync will cause havoc. A new release that does support what you want is just around the corner...

**Comment by charlestek** [comment]
You didn’t read what I wrote in the last email AT ALL.   I said that I had removed OutlookGoogleCalendar Sync from the 2nd PC many days ago.

This happened with ONE installation of OGCS on ONE PC.   Clearly having shut down outlook when OGCS ran screwed up something that caused it to delete all these appointments from both Calendars.

And NO, THE ITEMS WERE NOT in my Deleted ITEMS FOLDER IN OUTLOOK AT ALL.   The Google items were in Google Trash, but had to restore 66 entries one by one.

I am doing a system restore on the Windows installation.

And in all due respect to all the hard work you have done, having a graphic diagram of from, to and bidirectional does not give any more information than the drop downs in the OGCS.

The diagram is redundant.   The deletion interaction needs to be described in English words.    I used to be a .Net developer and your diagram makes no sense to me at all.

The precedence of flow needs to be described in English.  The fact that you understand it does not mean that other people understand it by a long shot.

I cannot use your program with the deficits of it deleting all my appointments when it decides to sync and fail when outlook is closed,  then upon outlook restart and manually sync it decides to delete all the appointments in its specified time span range.

I already  lost $50 due to a missed medical appointment with a no show fee.

There are other products on the market which you should install and compare as part of due diligence.

I suggest you look at https://www.makeuseof.com/tag/how-to-sync-microsoft-outlook-with-google-calendar/

In particular,  http://www.pppindia.com/calendar-sync/index.html

The fact that this gui tells you specifically which is master (an arrow does not mean the left hand side is a master)

And that you can select only one radio box choice at time, makes complete sense.    Apparently I have been running with your program with both calendars as a master which should not even be allowed.

In any case you have a fatal bug, if someone decides to close outlook when OGCS syncs.

**Comment by phw198** [comment]
No, I did read what you wrote and uninstalling one of the instances of OGCS doesn't magically remove the metadata stored against the calendar items in Google and Outlook due to two way sync. 

You've still worrying about the direction of sync arrows when the real problem is multiple installations of OGCS syncing to a common calendar. The common calendar can only hold metadata about **one** calendar from which it is syncing, not two - and this causes the metadata to continually change depending on which calendar last synced with it. With two way sync enabled, this can be interpreted as a delete, which isn't correct. The next version will support up to 99 different calendar syncing to the same common calendar, hence the problem being resolved. The advice up to now has always been to sync to/from an empty dedicated calendar.

With two way sync, three's no such thing as "master", only most recently updated. If it's one way sync, then obviously the source is master. Having recovered your Google items you _could_ have used OGCS with one way sync (Google-> Outlook, disable deletes) to repopulate Outlook, but hey, I'll leave you with pppindia. Sorry for any inconvenience OGCS may have caused!

**Comment by charlestek** [comment]
Paul,

Thank you for explaining this.

Unfortunately I do not know the internals of the program, nor would I be expected to.    So how do you remove the metadata from where it is stored…. Google Calendar ?  or Outlook calendar where the remaining instance of OGCS is running?   Or the other PC where OGCS was uninstalled?

**Comment by charlestek** [comment]
The sync arrows are a separate issue.   I am saying that the operation of the program is not clear, not clearly documented, and clearly not intuitive, at all in the settings section.   I have no idea what effect enabling both Outlook ->Google and Google <- Outlook together does, and I also wonder if this should be allowed, since I cannot intuit or figure out from the documentation what this does.

**Comment by charlestek** [comment]
“but hey, I'll leave you with pppindia”

I didn’t say pppdindia was great, in fact they require the google password to be stored in the program, which is a no-no, I don’t even know if they store it as a salted hash or in plain text.

What I did say was that their verbiage of the sync directions and options on their gui, (yellow highlighted box) IN MY OPINION is much clearer than the gui in OGCS.

Right now, and no reflection on your hard work for your application,   I cannot afford to have issues from an alpha or beta that results in loss of appointments.

It was my mistake to try to run OGCS on two PC’s.

I will have to try one of the other applications in the MakeUseOf article.

I should know better than not to have automatic backups of both the Outlook Calendar appointments and Google Calendar.    Unfortunately in my main PC, everything is backed up daily with Macrium, except for the fact that the PST for that machine is on another drive for a dual boot, and I did not set Macrium to back up that item.     This has pushed me to get that done.  I have to find something to auto backup the Google Calendar.

**Comment by phw198** [comment]
OK, given you've got/getting a good backup regime in place, and I'm literally hours away from finish testing a version that should work for what you need (and that's been months in the works), let's see if we can work on this together - I could really do with someone willing to use it in a "real world" setup ahead of general release :slightly_smiling_face: 

Firstly, those 61 deletions you reported would not have been from the Outlook crash, but due to the metadata I previously referenced.

> I have no idea what effect enabling both Outlook ->Google and Google <- Outlook together does

So this sounds like maybe Outlook at work (let's call it "OW") syncing one way to a Google calendar "G", and then another installation of OGCS and Outlook at home "OH" syncing one way to that very same Google calendar "G". That would be this diagram:
![image](https://user-images.githubusercontent.com/11843015/44307293-d264a300-a397-11e8-93db-42689f8cd50b.png)

In term of metadata for one-way sync, the destination calendar "G" only has one "slot" to record the appointment IDs from  the source calendar. That's fine for OW->G. Now we introduce OH->G and if you have the same items in OH, OGCS will see the OGCS metadata in G but it doesn't match so it will say "that needs deleting in G, it's not in the source (OH) calendar". If you have "merge" turned on or "deletions disabled", it will create a "duplicate" in G.

Now let's turn on two-way sync:
![image](https://user-images.githubusercontent.com/11843015/44307362-f379c380-a398-11e8-85ca-30891c27e25c.png)
Things get even more complicated. The OW->G step is the same as before, but it _also_ stores the details of the event in G it just created in OW (so when it sync G->OW it can find the match easily). The problem comes in again with G->OH, specifically with that back-population of metadata into G after creating the appointment in OH - it's just trampled the original metadata in that single "slot"in G! Now when OW<->G fires again, it doesn't match which equates to "that needs deleting". :confounded: 

In a nutshell, the solution is multiple metadata slots. And once this release is out, all of the above intricacies no longer need to be understood nor, IMO, explained in some form of succinct layman language.

>So how do you remove the metadata from where it is stored…. Google Calendar

Assuming OW is the remaining install of OGCS, turn this to one-way sync OW->G - steps are [here in the wiki](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Hints-&-Tips#fixing-a-stubborn-calendar-sync-with-full-sync). That full one-way sync will make sure all the meta data in G is correct.

**Comment by phw198** [comment]
New OGCS build available on #382 if you would like to try it.
